nuxeo-core/nuxeo-core-storage-sql/nuxeo-core-storage-sql/src/main/java/org/nuxeo/ecm/core/storage/sql/kv/SQLKeyValueStore.java

```diff
@@ -0,0 +1,1096 @@
+                            column.getFullQuotedName(), expected, actual, actualName, actualSize);
+                }
+            }
+        }
+    }
+
+    protected String getDatabaseSchemaName(Connection connection) throws SQLException {
+        String schemaName = null;
+        if (dialect instanceof DialectOracle) {
+            try (Statement st = connection.createStatement()) {
+                String sql = "SELECT SYS_CONTEXT('USERENV', 'SESSION_USER') FROM DUAL";
+                logger.log(sql);
+                try (ResultSet rs = st.executeQuery(sql)) {
+                    if (rs.next()) {
+                        schemaName = rs.getString(1);
+                        logger.log("  -> schema: " + schemaName);
+                    }
+                }
+            }
+        }
+        return schemaName;
+    }
+
+    protected void expireTTLOnce() {
+        runWithConnection(connection -> {
+            try {
+                try (PreparedStatement ps = connection.prepareStatement(expireSQL)) {
+                    Long ttlDeadline = getTTLValue(0);
+                    setToPreparedStatement(expireSQL, ps, ttlCol, ttlDeadline);
+                    int count = ps.executeUpdate();
+                    logger.logCount(count);
+                }
+            } catch (SQLException e) {
+                if (dialect.isConcurrentUpdateException(e)) {
+                    // ignore
+                    return;
+                }
+                log.debug("Exception during TTL expiration", e);
+            }
+        });
+    }
+
+    @Override
+    public void clear() {
+        runWithConnection(connection -> {
+            try (Statement st = connection.createStatement()) {
+                logger.log(deleteAllSQL);
+                st.execute(deleteAllSQL);
+            }
+        });
+    }
+
+    @Override
+    public Stream<String> keyStream() {
+        return runWithConnection((Connection connection) -> keyStream(connection, null));
+    }
+
+    @Override
+    public Stream<String> keyStream(String prefix) {
+        return runWithConnection((Connection connection) -> keyStream(connection, prefix));
+    }
+
+    protected Stream<String> keyStream(Connection connection, String prefix) throws SQLException {
+        String sql = prefix == null ? keyStreamSQL : keyStreamPrefixSQL;
+        List<String> keys = new ArrayList<>();
+        try (PreparedStatement ps = connection.prepareStatement(sql)) {
+            if (prefix != null) {
+                String like = escapeLike(prefix) + "%";
+                setToPreparedStatement(sql, ps, keyCol, like);
+            }
+            try (ResultSet rs = ps.executeQuery()) {
+                while (rs.next()) {
+                    String key = (String) keyCol.getFromResultSet(rs, 1);
+                    keys.add(key);
+                }
+            }
+        }
+        return keys.stream();
```

Comment:
Could be interesting to bind `ResultSet` retrieval directly into the `Stream` like it's done in mongo repository, wdyt ?

Reply:
We do that in `MongoDBRepository` for the internal `stream()` method but not currently in `MongoDBKeyValueStore.keyStream`. But yes it could be a nice improvement for MongoDB and SQL. I'll postpone this to another ticket 😄 
It also means that the caller must always close the stream properly otherwise we have a leak...